internal/enginex/enginex.go

```diff
@@ -0,0 +1,72 @@
+// Package enginex contains ooni/probe-engine extensions.
+package enginex
+
+import (
+	"encoding/json"
+
+	"github.com/fatih/color"
+	"github.com/ooni/probe-engine/log"
+	"github.com/ooni/probe-engine/model"
+)
+
+// LoggerAdapter exposes an interface compatible with the logger expected
+// by ooni/probe-engine, and forwards real log messages to a secondary logger
+// having the same interface (compatible with apex/log).
+type LoggerAdapter struct {
+	// Logger is the underlying logger
+	log.Logger
+}
+
+// Debug emits a debug message.
+func (la LoggerAdapter) Debug(msg string) {
+	la.Logger.Debug(color.WhiteString("engine") + ": " + msg)
+}
+
+// Debugf formats and emits a debug message.
+func (la LoggerAdapter) Debugf(format string, v ...interface{}) {
+	la.Logger.Debugf(color.WhiteString("engine")+": "+format, v...)
+}
+
+// Info emits an informational message.
+func (la LoggerAdapter) Info(msg string) {
+	la.Logger.Info(color.BlueString("engine") + ": " + msg)
+}
+
+// Infof format and emits an informational message.
+func (la LoggerAdapter) Infof(format string, v ...interface{}) {
+	la.Logger.Infof(color.BlueString("engine")+": "+format, v...)
+}
+
+// Warn emits a warning message.
+func (la LoggerAdapter) Warn(msg string) {
+	la.Logger.Warn(color.RedString("engine") + ": " + msg)
+}
+
+// Warnf formats and emits a warning message.
+func (la LoggerAdapter) Warnf(format string, v ...interface{}) {
+	la.Logger.Warnf(color.RedString("engine")+": "+format, v...)
+}
```

Comment:
Is there a reason to do the formatting of the log messages inside of here instead of inside of a apex/log handler: https://github.com/ooni/probe-cli/blob/master/internal/log/handlers/cli/cli.go

Reply:
The probe-engine does not depend on apex/log, rather it expects you to pass it [a structure that implements an interface that apex.log.Log happens to match](https://github.com/ooni/probe-engine/blob/79338ce6ff54e6497c9c4dfcb79d6f726269bcfc/log/log.go).

The pre-existing code was [filtering the logs emitted by MK and prefixing them with a coloured string](https://github.com/ooni/probe-cli/pull/46/files#diff-eacd810feae41332a4946c2af72b0b5aL118).

This change is my attempt at preserving the previously existing behaviour by making sure that messages emitted by probe-engine are still prefix.

I'd rather avoid depending on apex/log (and God forbids colouring messages) in the probe engine, but I think the following two questions are relevant to your discussion:

1. do we need to colour messages emitted by the engine during tests?

2. if the answer to the first question is affirmative, is there an apex/log way to construct an apex/log structure that (a) matches the required interface and (b) implements the desired behaviour that is more manageable than my implementation?

Reply:
tbh 1. is mostly an aestetic question. I think it would be nice, but I would not say we **need** it.

To do 2 inside of apex/log it's actually pretty simple (and as a matter of fact we already do it). When you call apex.log with a certain log level you get a `e.Level` which is the log level and you can color the message based on that, see: https://github.com/ooni/probe-cli/blob/master/internal/log/handlers/cli/cli.go#L134

Reply:
1. Right. I actually like this aesthetic feature.

2. Does this mean that there is a specific logger that I should pass probe engine here or that I can obtain the same result by constructing a new logger?

Reply:
> Does this mean that there is a specific logger that I should pass probe engine here or that I can obtain the same result by constructing a new logger?

 You just create any new instance of apex/log and then use it with it's log levels to obtain the desired result.

Maybe you want to give it an extra key to indicate it's coming from probe-engine. See as an example: https://github.com/ooni/probe-cli/blob/master/internal/output/output.go#L129

So you could do:

```
	log.WithFields(log.Fields{
		"type":            "engine_log",
	}).Info("Something")
```

Reply:
OK, so, as we just confirmed on Slack, I need to pass:

```Go
log.WithFields(log.Fields{
  "type": "engine",
})
```

to the probe-engine. This will give me a logger with fields. And then we can build on that to continue to have the coloured first entry as we used to have.